Python/context.c

```diff
@@ -257,6 +263,37 @@ PyContext_Exit(PyObject *octx)
 }
 
 
+void
+_PyContext_ExitThreadOwned(PyThreadState *ts)
+{
+    assert(ts != NULL);
+    // notify_context_watchers requires the notification to come from the
+    // affected thread, so we can only exit the context(s) if ts belongs to the
+    // current thread.
+    _Bool on_thread = ts == _PyThreadState_GET();
+    while (ts->context != NULL
+           && PyContext_CheckExact(ts->context)
+           && ((PyContext *)ts->context)->ctx_owned_by_thread
+           && on_thread) {
+        if (_PyContext_Exit(ts, ts->context)) {
+            Py_UNREACHABLE();
+        }
+    }
+    if (ts->context != NULL) {
+        // This intentionally does not use tstate variants of these functions
+        // (e.g., _PyErr_GetRaisedException(ts)) because ts might not belong to
+        // the current thread.
```

Comment:
That seems odd. If you pass a tstate to `_PyContext_ExitThreadOwned`, I would expect errors that occur in it to propagate to that thread state, rather than some arbitrary one.

Reply:
The purpose of the passed tstate is not to avoid a call to `PyThreadState_Get`, it's because that tstate is being destroyed.  The error should happen on the calling thread because that is the thread that is doing the destruction.  I updated the comments to hopefully make the code easier to understand.